sonar-ws/src/main/java/org/sonarqube/ws/client/WsClientFactory.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarqube.ws.client;
+
+import org.sonar.api.server.ws.LocalConnector;
+
+/**
+ * @since 5.5
+ */
+public class WsClientFactory {
```

Comment:
I was giving thoughs on what pattern we should recommend to plugin developer to follow and I hit an issue with having a non static `WsClientFactory`.

I thought of advising the plugin developer to design, let's say `FooWs`, to get injected with a `WsClientFactory` instance and add `WsClientFactory.getDefault()` to the component list returned by its `FooPlugin` class. This way, the UTs can use a test implementation of `WsClientFactory` and he/she can easily emulates internal Ws calls in SQ.

Now, assume the pattern described above hass been implemented: we design our components to be injected with `WsClientFactory` and therefor SQ's WebServer's container contains a `WsClientFactory.getDefault()` object.

Add to this the hypothesis that the plugin does not bundle the same version of the `WsClient` library as SQ does.

What is going to be Pico's behavior here? We have 2 objects of the same class `WsClientFactory` but from different class loaders?

I really don't know.

I'm starting to doubt the library approach...